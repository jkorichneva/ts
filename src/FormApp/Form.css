.form {
    height: 100%;
    width: 100%;
    padding: 60px 40px;
    background-color: white;
    transition: background-color 0.5s ease-out;
    box-sizing: border-box;
}

.form_theme-dark {
    background-color: #545454;
    color: #efefef;
}

.form .switcher {
    position: absolute;
    right: 50px;
    top: 20px;
}

.form .button {
    margin-top: 20px;
}
