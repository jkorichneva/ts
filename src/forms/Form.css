.form_theme-dark {
    background: #545454;
    color: #efefef;
}

.form {
    height: 100%;
    width: 100%;
    padding: 60px 40px;
}

.form .switcher {
    position: absolute;
    right: 50px;
    top: 20px;
}

.form .button {
    margin-top: 20px;
}
