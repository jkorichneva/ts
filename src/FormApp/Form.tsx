import React, {useState, useReducer, FormEvent} from 'react';
import Button from '../components/Button/Button';
import Input from '../components/Input/Input';
import Switcher from './fragments/Switcher/Switcher';
import Success from './fragments/Success/Success';
import initialState, {StateInt} from './constants/initialState';
import {getTheme, validate, validateForm} from "./utils/helpers";

import './Form.css';

type Action = ActionFieldChange | SimpleAction

type ActionFieldChange = {
    type: 'FIELD_CHANGE';
    event: React.FormEvent<HTMLInputElement>;
    field: string;
}

type SimpleAction = {
    type: 'SET_BUTTON_LOADING'|'SUBMIT';
}

function reducer(state: StateInt, action: Action): StateInt {
    switch (action.type) {
    case 'FIELD_CHANGE':
        /* tslint:disable-next-line: no-case-declarations */
        const validated = validate(action.field, action.event.currentTarget.value);
        if (validated) {
            return { ...state, [action.field]: action.event.currentTarget.value };
        } else {
            return state;
        }
    case 'SET_BUTTON_LOADING':
        return { ...state, buttonLoading: true };
    case 'SUBMIT':
        const validatedForm = validateForm(state);
        return { ...state, isCompleted: true };
    default:
        return state;
    }
}

function Form(): JSX.Element {
    const [isDark, setIsDark] = useState(false);
    const [{
        surname,
        name,
        job,
        phone,
        isCompleted,
        buttonLoading
    }, dispatch] = useReducer(reducer, initialState);
    return (
        <div className={`form form_theme-${isDark ? 'dark' : 'light'}`}>
            <Switcher isDark={isDark} onThemeChange={(): void => setIsDark(!isDark)}/>
            {!isCompleted && (<>
                <Input
                    type="text"
                    placeholder="Surname"
                    theme={getTheme(isDark)}
                    value={surname}
                    onChange={(event: FormEvent<HTMLInputElement>): void => dispatch({type: 'FIELD_CHANGE', event, field: 'surname'})}
                />
                <Input
                    type="text"
                    placeholder="Name"
                    theme={getTheme(isDark)}
                    value={name}
                    onChange={(event: FormEvent<HTMLInputElement>): void => dispatch({ type: 'FIELD_CHANGE', event, field: 'name' })}
                />
                <Input
                    type="text"
                    placeholder="Job"
                    theme={getTheme(isDark)}
                    value={job}
                    onChange={(event: FormEvent<HTMLInputElement>): void => dispatch({ type: 'FIELD_CHANGE', event, field: 'job' })}
                />
                <Input
                    type="tel"
                    placeholder="Phone number"
                    theme={getTheme(isDark)}
                    value={phone}
                    onChange={(event: FormEvent<HTMLInputElement>): void => dispatch({ type: 'FIELD_CHANGE', event, field: 'phone' })}
                />
                <Button text="Apply" theme={getTheme(isDark)} isLoading={buttonLoading} onClick={(): void => {
                    dispatch({ type: 'SET_BUTTON_LOADING' });
                    setTimeout(() => {
                        dispatch({ type: 'SUBMIT' });
                    }, 5000);
                }}/>
            </>)}
            {isCompleted && <Success />}
        </div>
    )
}

export default Form;
