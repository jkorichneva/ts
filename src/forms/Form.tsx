import React, {useState, useReducer, ChangeEvent} from 'react';
import Button from '../components/Button/Button';
import Input from '../components/Input/Input';
import Switcher from '../fragments/Switcher/Switcher';
import Success from '../fragments/Success/Success';
import {getTheme, validate} from "../helpers";

import './Form.css';

type State = {
    surname: string;
    name: string;
    job: string;
    phone: string;
    isCompleted: boolean;
    buttonLoading: boolean;
}

type Action = ActionFieldChange | SimpleAction

type ActionFieldChange = {
    type: 'FIELD_CHANGE';
    event: React.ChangeEvent<HTMLInputElement>;
    field: string;
}

type SimpleAction = {
    type: 'SET_BUTTON_LOADING'|'SUBMIT';
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
    case 'FIELD_CHANGE':
        /* tslint:disable-next-line: no-case-declarations */
        const validated = validate(action.field, action.event.target.value);
        if (validated) {
            return { ...state, [action.field]: action.event.target.value };
        } else {
            return state;
        }
    case 'SET_BUTTON_LOADING':
        return { ...state, buttonLoading: true };
    case 'SUBMIT':
        return { ...state, isCompleted: true };
    default:
        return state;
    }
}

const initialState = {
    surname: '',
    name: '',
    job: '',
    phone: '',
    buttonLoading: false,
    isCompleted: false,
}

function Form(): JSX.Element {
    const [isDark, setIsDark] = useState(false);
    const [{
        surname,
        name,
        job,
        phone,
        isCompleted,
        buttonLoading
    }, dispatch] = useReducer(reducer, initialState);
    return (
        <div className={`form form_theme-${isDark ? 'dark' : 'light'}`}>
            <Switcher isDark={isDark} onThemeChange={(): void => setIsDark(!isDark)}/>
            {!isCompleted && (<>
                <Input
                    type="text"
                    placeholder="Surname"
                    theme={getTheme(isDark)}
                    value={surname}
                    onChange={(event: ChangeEvent<HTMLInputElement>): void => dispatch({type: 'FIELD_CHANGE', event, field: 'surname'})}
                />
                <Input
                    type="text"
                    placeholder="Name"
                    theme={getTheme(isDark)}
                    value={name}
                    onChange={(event: ChangeEvent<HTMLInputElement>): void => dispatch({ type: 'FIELD_CHANGE', event, field: 'name' })}
                />
                <Input
                    type="text"
                    placeholder="Job"
                    theme={getTheme(isDark)}
                    value={job}
                    onChange={(event: ChangeEvent<HTMLInputElement>): void => dispatch({ type: 'FIELD_CHANGE', event, field: 'job' })}
                />
                <Input
                    type="tel"
                    placeholder="Phone number"
                    theme={getTheme(isDark)}
                    value={phone}
                    onChange={(event: ChangeEvent<HTMLInputElement>): void => dispatch({ type: 'FIELD_CHANGE', event, field: 'phone' })}
                />
                <Button text="Apply" theme={getTheme(isDark)} isLoading={buttonLoading} onClick={(): void => {
                    dispatch({ type: 'SET_BUTTON_LOADING' });
                    setTimeout(() => {
                        dispatch({ type: 'SUBMIT' });
                    }, 5000);
                }}/>
            </>)}
            {isCompleted && <Success />}
        </div>
    )
}

export default Form;
