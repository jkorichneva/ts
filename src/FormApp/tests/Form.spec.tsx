import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Form from '../Form';

let container: HTMLElement|null;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    document.body.removeChild(container);
    container = null;
});

test('renders Form', () => {
    act(() => {
        ReactDOM.render(<Form />, container);
    });
    expect(container).toMatchSnapshot();
});
