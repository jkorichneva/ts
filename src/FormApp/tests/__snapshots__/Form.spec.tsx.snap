// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Form 1`] = `
<div>
  <div
    class="form form_theme-light"
  >
    <div
      class="switcher"
    >
      <label
        class="switcher_switch"
      >
        <input
          type="checkbox"
        />
        <span
          class="switcher_slider"
        />
      </label>
      Dark Mode
    </div>
    <input
      class="input input_light input_size-m"
      placeholder="Surname"
      type="text"
      value=""
    />
    <input
      class="input input_light input_size-m"
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      class="input input_light input_size-m"
      placeholder="Job"
      type="text"
      value=""
    />
    <input
      class="input input_light input_size-m"
      placeholder="Phone number"
      type="tel"
      value=""
    />
    <button
      class="button button_light button_size-m"
    >
      Apply
    </button>
  </div>
</div>
`;
